kopykat-ksp/src/main/kotlin/at/kopyk/utils/TypeSubstitution.kt

```diff
@@ -0,0 +1,47 @@
+package at.kopyk.utils
+
+import com.google.devtools.ksp.symbol.KSClassDeclaration
+import com.google.devtools.ksp.symbol.KSName
+import com.google.devtools.ksp.symbol.KSReferenceElement
+import com.google.devtools.ksp.symbol.KSType
+import com.google.devtools.ksp.symbol.KSTypeAlias
+import com.google.devtools.ksp.symbol.KSTypeArgument
+import com.google.devtools.ksp.symbol.KSTypeParameter
+import com.google.devtools.ksp.symbol.KSTypeReference
+import com.google.devtools.ksp.symbol.Variance
+
+internal typealias TypeSubstitution = Map<KSName, KSType?>
+
+/* This function needs to "jump" across type aliases
+   to get the type parameters applied to the "real" type */
+internal fun KSTypeAlias.unravelTypeParameters(): TypeSubstitution {
+  val ty = type.resolve()
+  return when (val decl = ty.declaration) {
+    is KSClassDeclaration ->
+      decl.typeParameters.map { it.name }.zip(ty.arguments.map { it.type?.resolve() }).toMap()
```

Comment:
It required a bit of refactoring (because we cannot just `mapNotNull` the arguments, as they would lose sync with the parameter names), but it should be done in the next commit